plugins/heroku/heroku_log_input.go

```diff
@@ -0,0 +1,204 @@
+/***** BEGIN LICENSE BLOCK *****
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# The Initial Developer of the Original Code is the Mozilla Foundation.
+# Portions created by the Initial Developer are Copyright (C) 2014
+# the Initial Developer. All Rights Reserved.
+#
+# This plugin is based on the HttpListenPlugin.
+#
+# Contributor(s):
+#   Anton Lindström (carlantonlindstrom@gmail.com)
+#
+# ***** END LICENSE BLOCK *****/
+
+package heroku
+
+import (
+	"code.google.com/p/go-uuid/uuid"
+	"fmt"
+	"github.com/mozilla-services/heka/message"
+	"github.com/mozilla-services/heka/pipeline"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type HerokuLogInput struct {
+	conf        *HerokuLogInputConfig
+	listener    net.Listener
+	stopChan    chan bool
+	ir          pipeline.InputRunner
+	server      *http.Server
+	starterFunc func(hli *HerokuLogInput) error
+	hostname    string
+}
+
+const LOGPLEX_CONTENT_TYPE = "application/logplex-1"
+
+// Heroku Log Input config struct
+type HerokuLogInputConfig struct {
+	// TCP Address to listen to
+	// Defaults to "0.0.0.0:8325".
+	Address string
+}
+
+func (hli *HerokuLogInput) ConfigStruct() interface{} {
+	return &HerokuLogInputConfig{
+		Address: "0.0.0.0:8325",
+	}
+}
+
+func herokuStarter(hli *HerokuLogInput) (err error) {
+	hli.listener, err = net.Listen("tcp", hli.conf.Address)
+	if err != nil {
+		return fmt.Errorf("Listener [%s] start fail: %s\n",
+			hli.conf.Address, err.Error())
+	} else {
+		hli.ir.LogMessage(fmt.Sprintf("Listening on %s", hli.conf.Address))
+	}
+
+	err = hli.server.Serve(hli.listener)
+	if err != nil {
+		return fmt.Errorf("Serve fail: %s\n", err.Error())
+	}
+
+	return nil
+}
+
+// A Heroku log drain recieves messages on this format:
+var herokuLogRegexp = regexp.MustCompile(`^\d+ <(\d+)>(\d+) (.*) (.*) (.*) (.*) - (.*)$`)
+
+// Example:
+//   83 <40>1 2012-11-30T06:45:29+00:00 host app web.3 - State changed from starting to up
+
+func (hli *HerokuLogInput) RequestHandler(req *http.Request) error {
+	if req.Method != "POST" || len(req.URL.Path) < 2 ||
+		req.Header.Get("Content-Type") != LOGPLEX_CONTENT_TYPE {
+		return fmt.Errorf("Bad request")
+	}
+
+	msgCount, err := strconv.Atoi(req.Header.Get("Logplex-Msg-Count"))
+	if err != nil {
+		return fmt.Errorf("Could not parse Logplex-Msg-Count: %s", err.Error())
```

Comment:
Personally I think it would be good to create identificators like InvalidMsgCountError to easier handling later.